Autograd: Delay warnings until the end of backward execution

peterbell10 opened the pull request:
Fixes #50209

This adds a new warning handler that stores all warnings in a shared
queue, which can be "replayed" at a later time and, crucially, on
another thread. Then, I use this inside the autograd engine to ensure
that warnings are processed by the handler registered on the main
thread.

For testing, I also add an operator that always warns in the backward
pass and test that the warning is a normal Python warning.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

peterbell10 committed:
Autograd: Delay warnings until the end of backward execution

Fixes #50209

This adds a new warning handler that stores all warnings in a shared
queue, which can be "replayed" at a later time and, crucially, on
another thread. Then, I use this inside the autograd engine to ensure
that warnings are processed by the handler registered on the main
thread.

For testing, I also add an operator that always warns in the backward
pass and test that the warning is a normal Python warning.

peterbell10 committed:
WarningHandler: Remove virtual destructor

peterbell10 committed:
Assert a new warning handler is set when replaying warnings

albanD reviewed (approved):
Thanks for taking the time to redo this!

peterbell10 committed:
Restore WarningHandler virtual destructor

peterbell10 committed:
Move PyWarningHandler::process into a new class

This allows the class inheriting from c10::WarningHandler to have a
noexcept destructor.

albanD reviewed (approved):
Thanks
